### ISSUE TITLE
Allow upload of library headers

### ISSUE THREAD

**Issue by studoot** [mattgodbolt/compiler-explorer#405]
Further to #397 - it'd be nice to have the ability to upload header files (single files or maybe a zip file containing a tree of headers).

There are some obvious risks and complications to this that would need to be addressed (what size/number of header files could be allowed, upload of any file is a bit of a threat, how long do the files stick around for, if a session is revisited).

**Comment by mattgodbolt** [comment]
Absolutely. This has been on my personal TODO list for a while: thanks for making an issue to track it with! Lots needs to be thought about, as you say.

**Comment by danielmrussell** [comment]
This would be great. Another thing that would be very useful is simply the ability to cross-reference open source editors, without the need to upload. I suppose this would require that each source tab has a real name instead of the current generic "C++ source #1" etc naming scheme.

Currently, unless I'm missing something, it seems like we have to copy-paste the text from one into another in order to test them together. If there were a way to simply pull from existing open source editor tabs to use their stuff in another tab, that would be really cool.

Thanks for all Compiler Explorer does already though. It's awesome!

**Comment by mattgodbolt** [comment]
Cross-referencing editors is a thought. Each pane is very independent though. It would be a pretty big change to tie them together. But so would uploading header files etc, so this is a good thought!
